WFCORE-2620 - Add ability to read computed runtime values of IO subsystem buffer-pool attributes

rpelisse opened the pull request:
Issue: https://issues.jboss.org/browse/JBEAP-6984
Upstream Issue: https://issues.jboss.org/browse/WFCORE-2620
No upstream PR required

This changes is (hopefully) a better proposal than my previous changes. It is however just a draft - I don't expect it to be perfect, I'm sharing it to get feedback on what is wrong (and hopefully how to fix it). So please don't merge without thorough code review, thanks !

rpelisse commented:
@ctomc Please take a look when you have a sec' ! :)

bstansberry commented:
Default values in the management API description are static.

rpelisse commented:
@bstansberry I took a little week to reflect on your comment, but I'm really not sure what you mean here.   Would you mind elaborating ? :/ 

@ctomc btw, any thoughts from your side on this ?

bstansberry commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -85,36 +81,29 @@
             .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES)
             .setAllowExpression(true)
             .setValidator(new IntRangeValidator(1, true, true))
+            .setDefaultValue(new ModelNode(defaultBufferSize))
```
This isn't ok (same with the others). Only static values are allowed.

rpelisse commented:
@bstansberry thanks for the clarification - of course, now, it seems indeed very clear what was the issue :) - anyhow, I've fixed that, hopefully, in a good manner. Let me know if you see any other issue with this PR!

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolAdd.java:
```diff
@@ -0,0 +1,41 @@
+    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model)
+            throws OperationFailedException {
+        final PathAddress address = PathAddress.pathAddress(operation.get(OP_ADDR));
+        final String name = address.getLastElement().getValue();
```
You can directly use `org.jboss.as.controller.OperationContext#getCurrentAddressValue`

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -55,66 +51,66 @@
             .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES)
             .setAllowExpression(true)
             .setValidator(new IntRangeValidator(1, true, true))
+            .setDefaultValue(DEFAULT_BUFFER_SIZE)
```
You can not set default values for these attributes as they are derived from `Runtime.getRuntime().maxMemory()`.
Their actual values will depend on the VM they are running and will not be provide the same "default" across all servers in domain mode.

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -55,66 +51,66 @@
             .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES)
             .setAllowExpression(true)
             .setValidator(new IntRangeValidator(1, true, true))
+            .setDefaultValue(DEFAULT_BUFFERS_PER_REGION)
```
same as above

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -55,66 +51,66 @@
     static final SimpleAttributeDefinition DIRECT_BUFFERS = new SimpleAttributeDefinitionBuilder(Constants.DIRECT_BUFFERS, ModelType.BOOLEAN, true)
             .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES)
             .setAllowExpression(true)
+            .setDefaultValue(DEFAULT_DIRECT_BUFFERS)
```
same as above

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -55,66 +51,66 @@
-    private static final int defaultBufferSize;
-    private static final int defaultBuffersPerRegion;
-    private static final boolean defaultDirectBuffers;
+    static final int defaultBufferSize;
```
there is no need to store the primitive values as they are wrapped in the ModelNode fields

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -128,29 +124,29 @@ public void registerCapabilities(ManagementResourceRegistration resourceRegistra
-                    .install();
-
+        public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+            if ( definition.getType() == ModelType.INT )
```
you need here to find the name of the attribute to read and return one of DEFAULT_BUFFERS_PER_REGION, DEFAULT_BUFFER_SIZE or DEFAULT_DIRECT_BUFFERS

rpelisse commented:
@jmesnil thanks for the feedback! I've rework my implementation according to it (and also found a but at the same time). Let me know if you think I can still improve on this

rpelisse commented:
retest this please

rpelisse commented:
retest this please

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolAdd.java:
```diff
@@ -0,0 +1,39 @@
+        final boolean direct = directModel.isDefined() ? directModel.asBoolean()
+                : BufferPoolResourceDefinition.DEFAULT_DIRECT_BUFFERS.asBoolean();
+
+        System.out.println("BufferSize:" + bufferSize + ", bufferPerSlice:" + bufferPerSlice + ", direct:" + direct);
```
must be removed

rpelisse replied:
doh... sorry.

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolAdd.java:
```diff
@@ -0,0 +1,39 @@
+
+        System.out.println("BufferSize:" + bufferSize + ", bufferPerSlice:" + bufferPerSlice + ", direct:" + direct);
+        final BufferPoolService service = new BufferPoolService(bufferSize, bufferPerSlice, direct);
+        System.out.println("RPE>>>> name:" + name);
```
idem

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -42,79 +42,70 @@
-
-
-    /*<buffer-pool name="default" buffer-size="1024" buffers-per-slice="1024"/>*/
+    static final SimpleAttributeDefinition BUFFER_SIZE = new SimpleAttributeDefinitionBuilder(Constants.BUFFER_SIZE,
```
putting everything on the same line makes it hard to read and mantain.

Could you revert to the previous formatting, please?

rpelisse replied:
I think the Eclipse formatter screwed me on this one. I'll fix it indeed, thanks!

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -42,79 +42,70 @@
-    );
+    /* <buffer-pool name="default" buffer-size="1024" buffers-per-slice="1024"/> */
 
+    static List<SimpleAttributeDefinition> ATTRIBUTES = Arrays.asList(BUFFER_SIZE, BUFFER_PER_SLICE, DIRECT_BUFFERS);
```
idem

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -42,79 +42,70 @@
-                new BufferPoolAdd(),
-                new ReloadRequiredRemoveStepHandler()
-        );
+        super(IOExtension.BUFFER_POOL_PATH, IOExtension.getResolver(Constants.BUFFER_POOL), BufferPoolAdd.INSTANCE,
```
idem

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -128,29 +119,46 @@ public void registerCapabilities(ManagementResourceRegistration resourceRegistra
-        private BufferPoolAdd() {
-            super(BufferPoolResourceDefinition.ATTRIBUTES);
+        private ModelNode returnConfigIfDefinedOrDefaultValueFor(String attributeName, ModelNode config, ModelNode defaultValue) {
+            System.out.println("RPE>>> config item is: " + config);
```
must be removed

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -128,29 +119,46 @@ public void registerCapabilities(ManagementResourceRegistration resourceRegistra
-    private static class BufferPoolAdd extends AbstractAddStepHandler {
+    @Override
+    public void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+        resourceRegistration.registerReadOnlyAttribute(BUFFER_SIZE, new BufferReadAttributeHandler(BUFFER_SIZE));
```
are they read-only? They are defined in the model and the user can define values when the buffer-pool resource is added.

rpelisse replied:
Yes, the attributes should be read-only, @ctomc mentioned that on the [upstream issue](https://issues.jboss.org/browse/WFCORE-2620?focusedCommentId=13406312&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13406312) (if I understood correctly, of course).

jmesnil replied:
That's not what he said :)

I'm paraphrasing him as he said that there is no need to define a custom write operation handler as the buffer pool can not be dynamically resized.
However the attribute must still use a default write operation that will modify the model and put the server in reload-required (so that the new value will be only taken into account when the server is reloaded).

rpelisse replied:
Of course! I've jumped from "can't be change dynamically" to "readonly" but it's not the same. My bad. I'll change that too...

jmesnil commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolResourceDefinition.java:
```diff
@@ -128,29 +119,46 @@ public void registerCapabilities(ManagementResourceRegistration resourceRegistra
-            context.getCapabilityServiceTarget().addCapability(IO_POOL_RUNTIME_CAPABILITY, service)
-                    .setInitialMode(ServiceController.Mode.ACTIVE)
-                    .install();
+            final ModelNode config = Resource.Tools.readModel(context.readResourceFromRoot(address));
```
code seems a bit convoluted.

Why don't you just call the `definition.resolveModelAttribute(context, model)`?
If it is defined, returns the value, else, returns one of the DEFAULT_XXX that you could pass to the class when you instantiate it with the corresponding attribute definition?

rpelisse replied:
...because I did not see this method! Thanks for the tip!

rpelisse commented:
@jmesnil OK, I think I've fixed everything. I've cleaned up the code of the Reader, I've added a Writer and made any change require a reload. Let me see if you see anything more I can enhance!

rpelisse commented:
retest this please

rpelisse commented:
retest this please

rpelisse commented:
retest this please

rpelisse commented:
The tests failure seems unrelated. In the meantime...

@jmesnil  @bstansberry do you still have some changes request? I think I've covered all you mentioned but maybe I've missed something (or introduce an other problem :/ )

rpelisse commented:
Just for clarity purpose: I've rebased this change, but I still have to work on it (and won't before at least a couple of weeks). So please disregard this PR for now.

darranl commented:
Hello @rpelisse I am just looking to clean up some of the PRs in the queue, is there likely to be any progress on this one in the short term?  If not would it better to close this one and either reopen or resubmit once ready?

darranl commented on io/subsystem/src/main/java/org/wildfly/extension/io/BufferPoolAdd.java:
```diff
@@ -0,0 +1,83 @@
+package org.wildfly.extension.io;
```
When this is ready to be looked at this file is missing it's copyright header.

rpelisse commented:
Yes, let's close this! This is a bit old and I have no time right now to explore it again. If somehow it becomes important, let me know, otherwise I'll set up a new PR when I can.
